File: script/make/protocol/lua/CheatProtocol.lua
CheatProtocol = {}

function CheatProtocol.encode(offset, protocol, data)
    if protocol == 60001 then
        local table = {}

        return table
    elseif protocol == 60002 then
        local table = {}
        -- 命令
        table[offset] = string.pack(">s2", data)
        offset = offset + 1
        return table
    else
        error(string.format('unknown protocol define: %d', protocol))
    end
end

function CheatProtocol.decode(offset, protocol, bytes)
    if protocol == 60001 then
        -- 命令列表
        local data = {}
        local dataLength = string.unpack(">I2", bytes, offset)
        offset = offset + 2
        for dataIndex = 1, dataLength do
            -- 
            -- 描述
            local dataDataDescription = string.unpack(">s2", bytes, offset)
            offset = offset + 2 + string.len(dataDataDescription)
            -- 命令
            local dataDataCommand = string.unpack(">s2", bytes, offset)
            offset = offset + 2 + string.len(dataDataCommand)
            -- object
            local dataData = {description = dataDataDescription, command = dataDataCommand}
            data[dataIndex] = dataData
        end
        return data
    elseif protocol == 60002 then
        -- 结果
        local data = string.unpack(">s2", bytes, offset)
        offset = offset + 2 + string.len(data)
        return data
    else
        error(string.format('unknown protocol define: %d', protocol))
    end
end

File: script/make/protocol/cs/CheatProtocol.cs
public static class CheatProtocol
{
    public static void Encode(System.Text.Encoding encoding, System.IO.BinaryWriter writer, System.UInt16 protocol, System.Object data) 
    {
        switch (protocol) 
        {
            case 60001:
            {
                return;
            }
            case 60002:
            {
                // 命令
                var dataBytes = encoding.GetBytes((System.String)data);
                writer.Write(System.Net.IPAddress.HostToNetworkOrder((System.Int16)dataBytes.Length));
                writer.Write(dataBytes);
                return;
            }
            default:throw new System.ArgumentException(System.String.Format("unknown protocol define: {0}", protocol));
        }
    }

    public static System.Object Decode(System.Text.Encoding encoding, System.IO.BinaryReader reader, System.UInt16 protocol) 
    {
        switch (protocol) 
        {
            case 60001:
            {
                // 命令列表
                var dataLength = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                var data = new System.Collections.Generic.List<System.Object>(dataLength);
                while (dataLength-- > 0)
                {
                    // 
                    // 描述
                    var descriptionLength = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                    var description = encoding.GetString(reader.ReadBytes(descriptionLength));
                    // 命令
                    var commandLength = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                    var command = encoding.GetString(reader.ReadBytes(commandLength));
                    // object
                    var item = new System.Collections.Generic.Dictionary<System.String, System.Object>() {{"description", description}, {"command", command}};
                    // add
                    data.Add(item);
                }
                return data;
            }
            case 60002:
            {
                // 结果
                var dataLength = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                var data = encoding.GetString(reader.ReadBytes(dataLength));
                return data;
            }
            default:throw new System.ArgumentException(System.String.Format("unknown protocol define: {0}", protocol));
        }
    }
}